### PR TITLE
Make it possible to run the Browser Functional Tests from the command line

### PR DESCRIPTION
**Pull request by anurse**
This PR configures the `npm test` command in the `client-ts\FunctionalTests` to automatically launch a browser (headless chrome, by default) and run the functional tests, collecting the results and writing them out.

There are a few elements to this:
* We register a Jasmine Reporter in the Functional Tests that generates [Test-Anywhere Protocol (TAP)](https://testanything.org/tap-version-13-specification.html) output describing the test run as `<li>` elements in a hidden `<ul>` element (optionally made visible by the `displayTap` query string value)
* The `run-tests.js` script, uses the [`selenium-webdriver`](https://seleniumhq.github.io/selenium/docs/api/javascript/index.html) package. It:
  1. Launches Selenium WebDriver
  2. Launches the server
  3. Instructs WebDriver to open the page on the server that runs the tests in a browser (any browser supported by Selenium WebDriver)
  4. Scrapes the DOM, reading the hidden `<li>` values and write them to the console, until the tests complete
  5. Terminates the browser, server and WebDriver.
* Now, the `run-tests.js` script basically acts as a script that writes TAP output for our functional tests. We can pipe that through any TAP-compatible reporter (we use `faucet` here) to report output.

**IMPORTANT:** This is not yet integrated into our build, the first pass is just to make it runnable and provide the infrastructure.

Also, this mechanism should be able to automate BrowserStack/Sauce Labs/some other browser lab. We'll need to work out the mechanisms for deploying the server code somewhere, but we should be able to use this to move forward with that.

Examples:

**Successful test run (click to embiggen):**

[](https://gfycat.com/WatchfulHarmoniousAmazontreeboa)

**Failed test run (click to embiggen):**

[](https://gfycat.com/VillainousDeficientDuiker)

TODO
* [x] Rebase
* [x] Fix tslint errors (thanks @BrennanConroy :(... but seriously it's good)
* [x] Don't build the entire client in FunctionalTests.csproj, we can make sure it builds before we run the tests in the CI but doing it in FunctionalTests.csproj is bogging down VS builds.

### CODE REVIEW

**Comment by anurse** [comment]
Had to leave before tslint errors were all resolved but they're minimal and I will fix them before check-in (After all the build will be red ;))

File: client-ts/FunctionalTests/FunctionalTests.csproj [hunk 1]
```diff
@@ -26,11 +26,6 @@
   </ItemGroup>
 
   <Target Name="ClientBuild" BeforeTargets="AfterBuild">
-      <!-- This will result in a double build of the JavaScript modules. Not sure the best way to deal with that right now. -->
-      <Exec Command="npm run build" WorkingDirectory="$(MSBuildThisFileDirectory)/.." />
```
Comment (davidfowl) [inline client-ts/FunctionalTests/FunctionalTests.csproj]: YAAS

**Comment by anurse** [comment]
The TODOs are to be done in the PR before merge. In fact, they are done now.

File: client-ts/FunctionalTests/package.json [hunk 2]
```diff
@@ -6,14 +6,25 @@
...
+    "@types/selenium-webdriver": "^3.0.8",
+    "@types/yargs": "^11.0.0",
+    "es6-promise": "^4.2.2",
+    "faucet": "0.0.1",
+    "selenium-webdriver": "^4.0.0-alpha.1",
+    "tap-mocha-reporter": "^3.0.6",
+    "tree-kill": "^1.2.0",
+    "ts-node": "^4.1.0",
+    "webdriver-manager": "^12.0.6",
+    "yargs": "^11.0.0"
```
Comment (anurse) [inline client-ts/FunctionalTests/package.json]: FYI @Eilon, some additional dev-time-only dependencies.

Comment (Eilon) [reply client-ts/FunctionalTests/package.json]: Got it. I assume you did/will do the needful?

Comment (anurse) [reply client-ts/FunctionalTests/package.json]: Indeed I shall, just raising it FYI

**Comment by anurse** [comment]
Still got some flaky tests going on in AppVeyor... @BrennanConroy / @davidfowl are you guys aware of these?

**Comment by davidfowl** [comment]
I'm trying to fix the flakyness, I can't get my last PR to pass a single run.

**Pull request merged**